aptos-move/mvhashmap/src/versioned_code.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Aptos
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::types::{MVCodeError, MVCodeOutput, TxnIndex};
+use aptos_crypto::hash::{DefaultHasher, HashValue};
+use aptos_types::{
+    executable::{Executable, ExecutableDescriptor},
+    write_set::TransactionWrite,
+};
+use crossbeam::utils::CachePadded;
+use dashmap::DashMap;
+use std::{
+    collections::{btree_map::BTreeMap, HashMap},
+    hash::Hash,
+    sync::{
+        atomic::{AtomicUsize, Ordering},
+        Arc,
+    },
+};
+
+const FLAG_DONE: usize = 0;
+const FLAG_ESTIMATE: usize = 1;
+
+/// Every entry in shared multi-version data-structure has an "estimate" flag
+/// and some content.
+struct Entry<V: TransactionWrite> {
+    /// Used to mark the entry as a "write estimate". Even though the entry
+    /// lives inside the DashMap and the entry access will have barriers, we
+    /// still make the flag Atomic to provide acq/rel semantics on its own.
+    flag: AtomicUsize,
+
+    /// The contents of the module as produced by the VM (can be WriteOp based on a
+    /// blob or CompiledModule, but must satisfy TransactionWrite to be able to
+    /// generate the hash below.
+    module: Arc<V>,
+    /// The hash of the blob, used instead of incarnation for validation purposes,
+    /// and also for uniquely identifying associated executables. None implies
+    /// that the stored module (WriteOp) actually corresponded to a deletion.
+    maybe_hash: Option<HashValue>,
+}
+
+/// A VersionedValue internally contains a BTreeMap from indices of transactions
+/// that update the given access path alongside the corresponding entries.
+struct VersionedValue<V: TransactionWrite, X: Executable> {
+    versioned_map: BTreeMap<TxnIndex, CachePadded<Entry<V>>>,
+
+    /// Executable based on the storage version of the module.
+    base_executable: Option<Arc<X>>,
+    /// Executables corresponding to published versions of the module, based on hash.
+    executables: HashMap<HashValue, Arc<X>>,
```

Comment:
Why this is not part of the Entry?
Is this an optimization in case two transactions compute the same executable?
Or you only lazily compute it on a fetch?

Reply:
It is so we don't recompute the executable every time we need it (based on the same code), yes

Reply:
it's basically a cache of executables based on the module (code) cryptographic hash, or for the storage version (can't change)